File: app/src/main/java/com/amaze/filemanager/filesystem/otg/OtgAmazeFilesystem.java
package com.amaze.filemanager.filesystem.otg;

import android.content.ContentResolver;
import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.documentfile.provider.DocumentFile;

import com.amaze.filemanager.file_operations.filesystem.filetypes.AmazeFile;
import com.amaze.filemanager.file_operations.filesystem.filetypes.AmazeFilesystem;
import com.amaze.filemanager.file_operations.filesystem.filetypes.ContextProvider;
import com.amaze.filemanager.utils.OTGUtil;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;

import kotlin.NotImplementedError;

public class OtgAmazeFilesystem extends AmazeFilesystem {
  public static final String TAG = OtgAmazeFilesystem.class.getSimpleName();

  public static final OtgAmazeFilesystem INSTANCE = new OtgAmazeFilesystem();

  public static final String PREFIX = "otg:/";

  static {
    AmazeFile.addFilesystem(INSTANCE);
  }

  private OtgAmazeFilesystem() {}

  @Override
  public String getPrefix() {
    return PREFIX;
  }

  @Override
  public int prefixLength(@NonNull String path) {
    if (path.length() == 0) {
      throw new IllegalArgumentException(
              "This should never happen, all paths must start with OTG prefix");
    }

    return super.prefixLength(path);
  }

  @NonNull
  @Override
  public String normalize(@NonNull String path) {
    return simpleUnixNormalize(path);
  }

  @NonNull
  @Override
  public String resolve(String parent, String child) {
    final String prefix = parent.substring(0, prefixLength(parent));
    final String simplePathParent = parent.substring(prefixLength(parent));
    final String simplePathChild = child.substring(prefixLength(child));

    return prefix + basicUnixResolve(simplePathParent, simplePathChild);
  }

  @NonNull
  @Override
  public String getDefaultParent() {
    return PREFIX + getSeparator();
  }

  @Override
  public boolean isAbsolute(AmazeFile f) {
    return true; // Relative paths are not supported
  }

  @NonNull
  @Override
  public String resolve(AmazeFile f) {
    return f.getPath();
  }

  @NonNull
  @Override
  public String canonicalize(String path) throws IOException {
    return normalize(path);
  }

  public boolean exists(AmazeFile f, @NonNull ContextProvider contextProvider) {
    @Nullable
    final Context context = contextProvider.getContext();
    if(context == null) {
      return false;
    }

    return OTGUtil.getDocumentFile(f.getPath(), context, false) != null;
  }
  public boolean isFile(AmazeFile f, @NonNull ContextProvider contextProvider) {
    @Nullable
    final Context context = contextProvider.getContext();
    if(context == null) {
      return false;
    }

    return OTGUtil.getDocumentFile(f.getPath(), context, false).isFile();
  }
  public boolean isDirectory(AmazeFile f, @NonNull ContextProvider contextProvider) {
    @Nullable
    final Context context = contextProvider.getContext();
    if(context == null) {
      return false;
    }

    return OTGUtil.getDocumentFile(f.getPath(), context, false).isDirectory();
  }
  public boolean isHidden(AmazeFile f) {
    return false;// There doesn't seem to be hidden files for OTG
  }

  public boolean canExecute(AmazeFile f, @NonNull ContextProvider contextProvider) {
    throw new NotImplementedError(); // No way to check
  }
  public boolean canWrite(AmazeFile f, @NonNull ContextProvider contextProvider) {
    @Nullable
    final Context context = contextProvider.getContext();
    if(context == null) {
      return false;
    }

    return OTGUtil.getDocumentFile(f.getPath(), context, false).canWrite();
  }
  public boolean canRead(AmazeFile f, @NonNull ContextProvider contextProvider) {
    @Nullable
    final Context context = contextProvider.getContext();
    if(context == null) {
      return false;
    }

    return OTGUtil.getDocumentFile(f.getPath(), context, false).canRead();
  }
  public boolean canAccess(AmazeFile f, @NonNull ContextProvider contextProvider) {
    return exists(f, contextProvider); //If the system says it exists, we can access it
  }

  public boolean setExecutable(AmazeFile f, boolean enable, boolean owneronly) {
    return false; //TODO find way to set
  }
  public boolean setWritable(AmazeFile f, boolean enable, boolean owneronly) {
    return false; //TODO find way to set
  }
  public boolean setReadable(AmazeFile f, boolean enable, boolean owneronly) {
    return false; //TODO find way to set
  }

  @Override
  public long getLastModifiedTime(AmazeFile f) {
    return 0;
  }

  @Override
  public long getLength(AmazeFile f, @NonNull ContextProvider contextProvider) throws IOException {
    @Nullable
    final Context context = contextProvider.getContext();

    if(context == null) {
      throw new IOException("Error obtaining context for OTG");
    }

    return OTGUtil.getDocumentFile(f.getPath(), context, false).length();
  }

  @Override
  public boolean createFileExclusively(String pathname) throws IOException {
    return false;
  }

  @Override
  public boolean delete(AmazeFile f, @NonNull ContextProvider contextProvider) {
    @Nullable
    final Context context = contextProvider.getContext();

    if(context == null) {
      return false;
    }

    return OTGUtil.getDocumentFile(f.getPath(), context, false).delete();
  }

  @Nullable
  @Override
  public String[] list(AmazeFile f, @NonNull ContextProvider contextProvider) {
    @Nullable final Context context = contextProvider.getContext();
    if(context == null) {
      return null;
    }

    ArrayList<String> list = new ArrayList<>();
    DocumentFile rootUri = OTGUtil.getDocumentFile(f.getPath(), context, false);
    for (DocumentFile file : rootUri.listFiles()) {
      list.add(file.getUri().getPath());
    }

    return list.toArray(new String[0]);
  }

  @Nullable
  @Override
  public InputStream getInputStream(AmazeFile f, @NonNull ContextProvider contextProvider) {
    @Nullable
    final Context context = contextProvider.getContext();
    ContentResolver contentResolver = context.getContentResolver();
    DocumentFile documentSourceFile = OTGUtil.getDocumentFile(f.getPath(), context, false);
    try {
      return contentResolver.openInputStream(documentSourceFile.getUri());
    } catch (FileNotFoundException e) {
      Log.e(TAG, "Error obtaining input stream for OTG", e);
      return null;
    }
  }

  @Nullable
  @Override
  public OutputStream getOutputStream(AmazeFile f, @NonNull ContextProvider contextProvider) {
    @Nullable
            final Context context = contextProvider.getContext();
    ContentResolver contentResolver = context.getContentResolver();
    DocumentFile documentSourceFile = OTGUtil.getDocumentFile(f.getPath(), context, true);
    try {
      return contentResolver.openOutputStream(documentSourceFile.getUri());
    } catch (FileNotFoundException e) {
      Log.e(TAG, "Error output stream for OTG", e);
      return null;
    }
  }

  @Override
  public boolean createDirectory(AmazeFile f, @NonNull ContextProvider contextProvider) {
    if(f.exists(contextProvider)) {
      return true;
    }

    @Nullable
            final Context context = contextProvider.getContext();
    if(context == null) {
      return false;
    }

    final String parent = f.getParent();
    if(parent == null) {
      return false;
    }

    DocumentFile parentDirectory = OTGUtil.getDocumentFile(parent, context, true);


    if (parentDirectory.isDirectory()) {
      parentDirectory.createDirectory(f.getName());
      return true;
    }

    return false;
  }

  @Override
  public boolean rename(AmazeFile f1, AmazeFile f2, @NonNull ContextProvider contextProvider) {
    @Nullable
    final Context context = contextProvider.getContext();
    ContentResolver contentResolver = context.getContentResolver();
    DocumentFile documentSourceFile = OTGUtil.getDocumentFile(f1.getPath(), context, true);
    return documentSourceFile.renameTo(f2.getPath());
  }

  @Override
  public boolean setLastModifiedTime(AmazeFile f, long time) {
    throw new NotImplementedError();
  }

  @Override
  public boolean setReadOnly(AmazeFile f) {
    return false;
  }

  @Override
  public AmazeFile[] listRoots() {
    File[] roots = File.listRoots();
    AmazeFile[] amazeRoots = new AmazeFile[roots.length];

    for (int i = 0; i < roots.length; i++) {
      amazeRoots[i] = new AmazeFile(roots[i].getPath());
    }

    return new AmazeFile[] { new AmazeFile(getDefaultParent()) };
  }

  @Override
  public long getTotalSpace(AmazeFile f, @NonNull ContextProvider contextProvider) {
    @Nullable
    final Context context = contextProvider.getContext();
    // TODO: Find total storage space of OTG when {@link DocumentFile} API adds support
    DocumentFile documentFile = OTGUtil.getDocumentFile(f.getPath(), context, false);
    return documentFile.length();
  }

  @Override
  public long getFreeSpace(AmazeFile f) {
    return 0; //TODO
  }

  @Override
  public long getUsableSpace(AmazeFile f) {
    // TODO: Get free space from OTG when {@link DocumentFile} API adds support
    return 0;
  }


}

File: app/src/main/java/com/amaze/filemanager/filesystem/otg/OtgAmazeFilesystem.kt
package com.amaze.filemanager.filesystem.otg

import android.util.Log
import com.amaze.filemanager.utils.OTGUtil.getDocumentFile
import com.amaze.filemanager.file_operations.filesystem.filetypes.AmazeFilesystem
import com.amaze.filemanager.file_operations.filesystem.filetypes.AmazeFile
import com.amaze.filemanager.file_operations.filesystem.filetypes.ContextProvider
import java.io.*
import java.util.ArrayList

class OtgAmazeFilesystem private constructor() : AmazeFilesystem() {
    companion object {
        val TAG = OtgAmazeFilesystem::class.java.simpleName

        val INSTANCE = OtgAmazeFilesystem()

        const val PREFIX = "otg:/"

        init {
            AmazeFile.addFilesystem(INSTANCE)
        }
    }

    override val prefix = PREFIX

    override fun prefixLength(path: String): Int {
        require(path.length != 0) { "This should never happen, all paths must start with OTG prefix" }
        return super.prefixLength(path)
    }

    override fun normalize(path: String): String {
        return simpleUnixNormalize(path)
    }

    override fun resolve(parent: String?, child: String?): String {
        val prefix = parent!!.substring(0, prefixLength(parent))
        val simplePathParent = parent.substring(prefixLength(parent))
        val simplePathChild = child!!.substring(prefixLength(child))
        return prefix + basicUnixResolve(simplePathParent, simplePathChild)
    }

    override val defaultParent: String
        get() = prefix + getSeparator()

    override fun isAbsolute(f: AmazeFile): Boolean {
        return true // Relative paths are not supported
    }

    override fun resolve(f: AmazeFile): String {
        return f.path
    }

    @Throws(IOException::class)
    override fun canonicalize(path: String?): String {
        return normalize(path!!)
    }

    override fun exists(f: AmazeFile, contextProvider: ContextProvider): Boolean {
        val context = contextProvider.getContext() ?: return false
        return getDocumentFile(f.path, context, false) != null
    }

    override fun isFile(f: AmazeFile, contextProvider: ContextProvider): Boolean {
        val context = contextProvider.getContext() ?: return false
        return getDocumentFile(f.path, context, false)!!.isFile
    }

    override fun isDirectory(f: AmazeFile, contextProvider: ContextProvider): Boolean {
        val context = contextProvider.getContext() ?: return false
        return getDocumentFile(f.path, context, false)!!.isDirectory
    }

    override fun isHidden(f: AmazeFile): Boolean {
        return false // There doesn't seem to be hidden files for OTG
    }

    override fun canExecute(f: AmazeFile, contextProvider: ContextProvider): Boolean {
        throw NotImplementedError() // No way to check
    }

    override fun canWrite(f: AmazeFile, contextProvider: ContextProvider): Boolean {
        val context = contextProvider.getContext() ?: return false
        return getDocumentFile(f.path, context, false)!!.canWrite()
    }

    override fun canRead(f: AmazeFile, contextProvider: ContextProvider): Boolean {
        val context = contextProvider.getContext() ?: return false
        return getDocumentFile(f.path, context, false)!!.canRead()
    }

    override fun canAccess(f: AmazeFile, contextProvider: ContextProvider): Boolean {
        return exists(f, contextProvider) //If the system says it exists, we can access it
    }

    override fun setExecutable(f: AmazeFile, enable: Boolean, owneronly: Boolean): Boolean {
        return false //TODO find way to set
    }

    override fun setWritable(f: AmazeFile, enable: Boolean, owneronly: Boolean): Boolean {
        return false //TODO find way to set
    }

    override fun setReadable(f: AmazeFile, enable: Boolean, owneronly: Boolean): Boolean {
        return false //TODO find way to set
    }

    override fun getLastModifiedTime(f: AmazeFile): Long {
        return 0
    }

    @Throws(IOException::class)
    override fun getLength(f: AmazeFile, contextProvider: ContextProvider): Long {
        val context = contextProvider.getContext()
                ?: throw IOException("Error obtaining context for OTG")
        return getDocumentFile(f.path, context, false)!!.length()
    }

    @Throws(IOException::class)
    override fun createFileExclusively(pathname: String?): Boolean {
        return false
    }

    override fun delete(f: AmazeFile, contextProvider: ContextProvider): Boolean {
        val context = contextProvider.getContext() ?: return false
        return getDocumentFile(f.path, context, false)!!.delete()
    }

    override fun list(f: AmazeFile, contextProvider: ContextProvider): Array<String>? {
        val context = contextProvider.getContext() ?: return null
        val list = ArrayList<String>()
        val rootUri = getDocumentFile(f.path, context, false) ?: return null
        for (file in rootUri.listFiles()) {
            file.uri.path?.let { list.add(it) }
        }
        return list.toTypedArray()
    }

    override fun getInputStream(f: AmazeFile, contextProvider: ContextProvider): InputStream? {
        val context = contextProvider.getContext()
        val contentResolver = context!!.contentResolver
        val documentSourceFile = getDocumentFile(f.path, context, false)
        return try {
            contentResolver.openInputStream(documentSourceFile!!.uri)
        } catch (e: FileNotFoundException) {
            Log.e(TAG, "Error obtaining input stream for OTG", e)
            null
        }
    }

    override fun getOutputStream(f: AmazeFile, contextProvider: ContextProvider): OutputStream? {
        val context = contextProvider.getContext()
        val contentResolver = context!!.contentResolver
        val documentSourceFile = getDocumentFile(f.path, context, true)
        return try {
            contentResolver.openOutputStream(documentSourceFile!!.uri)
        } catch (e: FileNotFoundException) {
            Log.e(TAG, "Error output stream for OTG", e)
            null
        }
    }

    override fun createDirectory(f: AmazeFile, contextProvider: ContextProvider): Boolean {
        if (f.exists(contextProvider)) {
            return true
        }
        val context = contextProvider.getContext() ?: return false
        val parent = f.parent ?: return false
        val parentDirectory = getDocumentFile(parent, context, true)
        if (parentDirectory!!.isDirectory) {
            parentDirectory.createDirectory(f.name)
            return true
        }
        return false
    }

    override fun rename(f1: AmazeFile, f2: AmazeFile, contextProvider: ContextProvider): Boolean {
        val context = contextProvider.getContext()
        val contentResolver = context!!.contentResolver
        val documentSourceFile = getDocumentFile(f1.path, context, true)
        return documentSourceFile!!.renameTo(f2.path)
    }

    override fun setLastModifiedTime(f: AmazeFile, time: Long): Boolean {
        throw NotImplementedError()
    }

    override fun setReadOnly(f: AmazeFile): Boolean {
        return false
    }

    override fun listRoots(): Array<AmazeFile> {
        val roots = File.listRoots()
        val amazeRoots = arrayOfNulls<AmazeFile>(roots.size)
        for (i in roots.indices) {
            amazeRoots[i] = AmazeFile(roots[i].path)
        }
        return arrayOf(AmazeFile(defaultParent))
    }

    override fun getTotalSpace(f: AmazeFile, contextProvider: ContextProvider): Long {
        val context = contextProvider.getContext()
        // TODO: Find total storage space of OTG when {@link DocumentFile} API adds support
        val documentFile = getDocumentFile(f.path, context!!, false)
        return documentFile!!.length()
    }

    override fun getFreeSpace(f: AmazeFile): Long {
        return 0 //TODO
    }

    override fun getUsableSpace(f: AmazeFile): Long {
        // TODO: Get free space from OTG when {@link DocumentFile} API adds support
        return 0
    }
}